### PR TITLE
Sparse checkout - kcl mod add --package 

### PR DESCRIPTION
**Pull request by officialasishkumar**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [ ] Y 

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Other

This PR adds a functionality that allows to add a single package in the dependency incase the repo contains multiple of them. It does this as follows: 

1. `kcl mod add --git <url> --commit <hash> --package <package_name>` (done)
    
After running this command, KPM will download the whole repo (also works if there is no kcl.mod in the root) and specifies package in the kcl.mod file. It changes the full_name, name and other similar variables in kcl.mod.lock and kcl.mod as per the package name (it doesn't uses the repo name if package flag is there). 

2. `kcl mod run`  (In progress)

This command will first checkout the directory which contains the package by running a recursive function. It will then do all the usual kcl compilation based on that directory. 
    
#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [ ] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [ ] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

### CODE REVIEW

**Comment by officialasishkumar** [comment]
PTAL @zong-zhe @Peefy 

This PR is different from the last one since here I am using the package functionality in add subcommand. I have written a description which explains more about this PR. I hope it helps!

I will update the `kcl mod run` command soon.

**Comment by zong-zhe** [comment]
Hi @officialasishkumar 😄 

There is a null pointer exception in the e2e test, you may need to check these test cases and fix the null pointer.

**Comment by officialasishkumar** [comment]
Hi @zong-zhe , 

Thank you for the comment. Will fix it soon!

**Comment by officialasishkumar** [comment]
@zong-zhe fixed the failing tests

File: pkg/client/client.go [hunk 1]
```diff
@@ -52,6 +52,8 @@ type KpmClient struct {
 	settings settings.Settings
 	// The flag of whether to check the checksum of the package and update kcl.mod.lock.
 	noSumCheck bool
+	// The package to use in case of multiple packages.
+	pkg string
```
Comment (zong-zhe) [inline pkg/client/client.go:56]: It may not be appropriate to add this field to KpmClient, which is the client of the package management tool, and it should not be associated with a specific package.

**Comment by zong-zhe** [comment]
Hi @officialasishkumar 😄 

Good job. The overall direction of your work is correct, and most of it has been completed. Now, there are still a few problems. In addition to the comments I gave you in the code, you also need to add some test cases for the functions you developed to ensure that your functions are normal. It mainly includes: 1. To test whether kcl.mod with `package` field can work well. 2. Test whether you can download a subpackage from a github repository.

File: pkg/client/client.go (hunk 1 above)
Comment (officialasishkumar) [reply pkg/client/client.go]: Thanks for the comment @zong-zhe. I have removed it!

**Comment by officialasishkumar** [comment]
Hey @zong-zhe,

I have addressed the comments. Please have a look!

**Comment by zong-zhe** [comment]
> Hey @zong-zhe,
> 
> I have addressed the comments. Please have a look!

Hi @officialasishkumar 😄 

Add a test case to test the dependency is added as expected to kcl.mod and kcl.mod.lock, prepare the expected.mod and expected.mod.lock, empty kcl.mod and kcl.mod.lock, after adding the dependency, compare the kcl.mod/kcl.mod.lock with expected.mod/expected.mod.lock.

**Comment by officialasishkumar** [comment]
Hi @zong-zhe 

i have added the mentioned test case

**Comment by officialasishkumar** [comment]
@Peefy Are there any other deliverables left for LFX?
